sql/grant_tables/privileged_database_provider.go

```diff
@@ -0,0 +1,336 @@
+// Copyright 2022 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grant_tables
+
+import (
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+// PrivilegedDatabaseProvider is a wrapper around a normal sql.DatabaseProvider that takes a context's client's
+// privileges into consideration when returning a sql.Database. In addition, any returned databases are wrapped with
+// PrivilegedDatabase.
+type PrivilegedDatabaseProvider struct {
```

Comment:
The problem with this approach is that you have to implement the entire DB interface, which we keep adding things to. It also adds an additional layer of object to the object graph, which is already pretty crowded.

Did you consider implementing these permission checks as part of an analyzer pass? I can't think off hand why that wouldn't work.

I don't want to slow down your work on this, but if that's an option I would reach for it.

Reply:
We talked about this directly but putting this here just so the information is public. The preferred way to do this would be to make this a part of the analyzer pass, rather than wrapping the `DatabaseProvider` and such. However, we are not exactly principled in that not all of our table acquisitions occur through the analyzer. Some grab the database from the analyzer, and then fetch the tables from that returned database outside of the analyzer. In addition, there are no restrictions in place to prevent such usage, so future statements are free to implement their database and table acquisition in any way.

To ensure that all statements are caught, I'm wrapping all possible access to databases and tables. In the future, when more time is available, the interfaces should be rewritten so that only the analyzer may request databases AND tables (potentially by removing the ability to fetch tables from a database itself), and this wrapper may be removed as the functionality will fully reside in the analyzer.